laurent22/joplin
# Desktop: Accessibility: Add ARIA information to the sidebar's notebook and tag list

# Summary

This pull request:

-   **Main goal**: Adds appropriate aria information to the sidebar (see [WCAG SC 1.3.1](https://www.w3.org/WAI/WCAG22/Understanding/info-and-relationships), [WCAG SC 1.3.6](https://www.w3.org/WAI/WCAG22/Understanding/identify-purpose)):
    -   [role=tree](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/tree_role) is used for the sidebar tag+notebook container.
    -  [role=treeitem](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/treeitem_role) is used for individual items in the tree.
    - [aria-level](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-level) is used to convey notebook depth. The NOTEBOOKS heading has `aria-level=1`, a toplevel notebook has `aria-level=2`, a notebook with one parent notebook has `aria-level=3`.
     -  Among other things, this causes NVDA, VoiceOver, and Orca to announce the depth of a notebook.

Additionally:

-   **Keyboard accessibility**: Makes it possible to expand/collapse the "TAGS"/"NOTEBOOKS" headings with the keyboard.
     -  See [WCAG SC 2.1.1](https://www.w3.org/TR/WCAG22/#keyboard).
-   **Side-effect of refactoring**: Fixes an issue where the :focusElementSidebar command did nothing if both "NOTEBOOKS" and "TAGS" were collapsed.
-   **Side-effect of refactoring**: Provides screen readers with information about the currently selected item in the sidebar, even if offscreen.
-   **Side-effect of refactoring:** Fixes an issue where the "NOTEBOOKS" and "TAGS" headers were slightly taller than expected by the note list container.

## Remaining issues & future work

-   Skip spacer items when navigating with the keyboard.
-   Fix "NOTEBOOKS" must be reselected after toggling it.
-   Remove tab override on the sidebar and note list.
-   (Optional) Improve focus-visible styling.

# Testing plan

1.   Enable the screen reader.
2.   Focus "All Notes"
3.   Press the down arrow until the last notebook has focus.
4.   Move focus to a notebook that has subnotebooks. Expand it if it isn't already expanded.
5.   Toggle the notebook using space.
6.   Toggle the notebook using the arrow keys.
7.   Focus the "NOTEBOOKS" header.
8. Hide, then show, all notebooks by pressing space.
9.   Hide, then show, all tags using the left/right arrow keys.
10.   Move focus to the "NOTEBOOKS" header.
11.   Open the context menu and create a new notebook using the context menu key.
12.   Make one notebook a subnotebook of another using the context menu key.
13.   Disable the screen reader.
14.   Drag a note from one notebook to another. Verify that the note is moved.

## Screen recording comparison: NVDA + Windows



[Screencast: Shows the above steps](https://github.com/user-attachments/assets/67d457ef-44ac-43ba-a8cf-7e62a61633fa)

<details><summary>Compare with before</summary>

Note: Previously, NVDA interacted with the note list in [browse mode](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#BrowseMode), due to its lack of a role. As such, NVDA handles the focus, rather than Joplin.

[Screencast: Shows the above steps done with a version of Joplin from before this pull request](https://github.com/user-attachments/assets/9164f6be-efa6-4036-835d-c16ac49fc6ec)

</details>
